### PR TITLE
added ability to ping marathon and to delete the leader

### PR DESCRIPTION
**Pull request by kensipe**

### CODE REVIEW

**Comment by kensipe** [comment]
@mesosphere-ci retest this please

**Comment by kensipe** [comment]
@mesosphere-ci retest this please

**Comment by kensipe** [comment]
it appears that the retest only retests the windows-binary :(

**Comment by tamarrow** [comment]
@kensipe thats because windows is the only CI using jenkins :)

**Comment by kensipe** [comment]
@tamarrow ready to go! 

The following are now true:

commands added:
```
dcos node dns <dns-name> [--json]
dcos marathon leader show
dcos marathon leader delete
dcos marathon ping [--once][--timeout=20]
```

```$ dcos marathon ping
Marathon ping response[3x]: pong```

```$ dcos marathon ping --once
Marathon ping response[1x]: pong
```

show marathon leader: 
```$ dcos marathon leader show
HOST                 IP      
marathon.mesos.  10.0.7.193 
```

then a delete:
```$ dcos marathon leader delete
Leadership abdicated
```

then a ping:  (the default time of 10s isn't long enough to usually get back a good response.)
```
$ dcos marathon ping --timeout 30
Marathon ping response[3x]: pong
```

the next dcos show should show a new leader (requires multi-master)
```
$ dcos marathon leader show
HOST                 IP      
marathon.mesos.  10.0.4.207  
```

The response if you timeout on ping:
```$ dcos marathon ping
Marathon ping response[3x]: Unable to get ping for all 3 masters in 10s
```

File: cli/dcoscli/marathon/main.py [hunk 1]
```diff
@@ -660,6 +677,89 @@ def group_scale(self, group_id, scale_factor, force):
...
+
+        :param json_: output json if True
+        :type json_: bool
+        :returns: process return code
+        :rtype: int
+        """
+
+        hosts = mesos.MesosDNSClient().hosts('marathon.mesos')
+        for leader in hosts:
+            pass
```
Comment (tamarrow) [inline cli/dcoscli/marathon/main.py]: why do we need this loop?

File: cli/dcoscli/marathon/main.py [hunk 2]
```diff
@@ -660,6 +677,89 @@ def group_scale(self, group_id, scale_factor, force):
...
+        if json_:
+            emitter.publish(leader)
```
Comment (tamarrow) [inline cli/dcoscli/marathon/main.py]: leader is not defined

File: cli/dcoscli/marathon/main.py [hunk 3]
```diff
@@ -660,6 +677,89 @@ def group_scale(self, group_id, scale_factor, force):
...
+        else:
+            table = tables.dns_table(hosts)
+            output = six.text_type(table)
```
Comment (tamarrow) [inline cli/dcoscli/marathon/main.py]: now that we only support python3, i don't think we ned the `text_type`

File: cli/dcoscli/data/help/marathon.txt [hunk 4]
```diff
@@ -28,6 +28,9 @@ Usage:
     dcos marathon group show [--group-version=<group-version>] <group-id>
     dcos marathon group remove [--force] <group-id>
     dcos marathon group update [--force] <group-id> [<properties>...]
+    dcos marathon leader delete
+    dcos marathon leader show [--json]
+    dcos marathon ping [--once] [--timeout=<timeout>]
```
Comment (tamarrow) [inline cli/dcoscli/data/help/marathon.txt]: having timeout + once seems a little redundant. There is also a global timeout param for the CLI. I would just keep the `once` param.

File: cli/dcoscli/marathon/main.py [hunk 5]
```diff
@@ -660,6 +677,89 @@ def group_scale(self, group_id, scale_factor, force):
...
+
+        if not timeout:
+            timeout = 10
+        else:
+            timeout = int(timeout)
+
+        consecutive_count = 0
+        client = self._create_marathon_client()
+
+        @retrying.retry(stop_max_delay=timeout*1000)
```
Comment (tamarrow) [inline cli/dcoscli/marathon/main.py]: instead of adding another dependency, can't we just loop + sleep?

File: cli/dcoscli/marathon/main.py [hunk 6]
```diff
@@ -660,6 +677,89 @@ def group_scale(self, group_id, scale_factor, force):
...
+            nonlocal consecutive_count
+            try:
+                response = client.ping()
+                consecutive_count = consecutive_count + 1
+            except:
+                consecutive_count = 0
+
+            if consecutive_count < count:
+                msg = 'Consecutive Count: {} is less than master count: {}'
+                raise Exception(msg.format(consecutive_count, count))
```
Comment (tamarrow) [inline cli/dcoscli/marathon/main.py]: we should only raise `DCOSException`

File: cli/dcoscli/marathon/main.py [hunk 7]
```diff
@@ -660,6 +677,89 @@ def group_scale(self, group_id, scale_factor, force):
...
+        def ping_for_each_master():
+            nonlocal consecutive_count
+            try:
+                response = client.ping()
+                consecutive_count = consecutive_count + 1
+            except:
+                consecutive_count = 0
+
+            if consecutive_count < count:
+                msg = 'Consecutive Count: {} is less than master count: {}'
```
Comment (tamarrow) [inline cli/dcoscli/marathon/main.py]: As a user I don't know what 'consecutive count' means

File: dcos/mesos.py [hunk 8]
```diff
@@ -315,6 +315,22 @@ def hosts(self, host):
         url = self._path('v1/hosts/{}'.format(host))
         return http.get(url, headers={}).json()
 
+    def masters(self):
```
Comment (tamarrow) [inline dcos/mesos.py]: 👍

File: cli/dcoscli/node/main.py [hunk 9]
```diff
@@ -438,6 +443,30 @@ def _bundle_create(nodes):
     return 0
 
 
+def _dns_lookup(dns_name, json_):
```
Comment (tamarrow) [inline cli/dcoscli/node/main.py]: can you add a table unit test for this please

**Comment by tamarrow** [comment]
bump

**Comment by jimenez** [comment]
ping @kensipe ?

**Comment by kensipe** [comment]
@jimenez thanks!   It seems that the email notifications for these pings aren't getting to me.. I will figure that out... I will work on this.. this week... probably today.

**Comment by kensipe** [comment]
rebased... addressing PR feedback today

File: cli/dcoscli/marathon/main.py (hunk 1 above)
Comment (kensipe) [inline cli/dcoscli/marathon/main.py]: not needed

File: cli/dcoscli/marathon/main.py (hunk 2 above)
Comment (kensipe) [inline cli/dcoscli/marathon/main.py]: it was defined... but there is only ever 1 value for marathon.mesos

File: cli/dcoscli/marathon/main.py (hunk 3 above)
Comment (kensipe) [inline cli/dcoscli/marathon/main.py]: roger that

File: cli/dcoscli/marathon/main.py (hunk 6 above)
Comment (kensipe) [inline cli/dcoscli/marathon/main.py]: +1

**Comment by kensipe** [comment]
rebased

**Comment by kensipe** [comment]
rebased

**Comment by kensipe** [comment]
rebased

**Comment by kensipe** [comment]
@mesosphere-ci retest this please

**Comment by kensipe** [comment]
rebased

**Comment by kensipe** [comment]
run linux integration tests

**Comment by kensipe** [comment]
@bamarni this was previous green... I rebased again.. anything that needs addressing prior to merge (other than waiting on tests again?)

**Comment by bamarni** [comment]
@kensipe : I haven't checked this PR yet, I'll try to pull and play with it tomorrow. However integration tests would be appreciated as this brings quite some new commands.

As for rebasing, I've just merged #1061, so there is no need to rebase anymore as long as there is no conflict (Jenkins merges with master before running tests).

**Comment by kensipe** [comment]
run windows tests

**Comment by kensipe** [comment]
run linux integration tests

**Comment by kensipe** [comment]
@bamarni integration tests added for new functionality

I debated adding in the "dcos marathon leader delete"... it will take out marathon for a short period of time.   I added it in and have a retry going on up to 5 mins... my experiences is that typically this is around 20-30 secs.   The retry logic will return sooner if it can.

**Comment by kensipe** [comment]
run linux integration tests

**Comment by kensipe** [comment]
run linux integration tests

**Comment by kensipe** [comment]
run linux integration tests

**Comment by kensipe** [comment]
run linux integration tests

**Comment by kensipe** [comment]
run linux integration tests

**Comment by kensipe** [comment]
run linux integration tests

**Comment by kensipe** [comment]
run linux integration tests

**Comment by kensipe** [comment]
YAY!   ready to merge @bamarni !

File: cli/tox.ini [hunk 10]
```diff
@@ -13,6 +13,7 @@ deps =
   pytest-cov
   pytz
   teamcity-messages
+  retrying
```
Comment (bamarni) [inline cli/tox.ini]: I'm really confused by this, so do we have 3 places to define dependencies for a package? Ideally we should have one place for dependencies and another one for dev / test dependencies. I've opened this ticket : https://jira.mesosphere.com/browse/DCOS_OSS-1762

But this should be tackled separately.

File: setup.py [hunk 11]
```diff
@@ -69,6 +69,7 @@
         'six>=1.9, <2.0',
         'toml>=0.9, <1.0',
         'sseclient==0.0.14',
+        'retrying==1.3.3',
```
Comment (bamarni) [inline setup.py]: also here.. this dependency is not needed for the dcos package.

**Pull request merged**